File: Python3/201 - 400/377 Combination Sum IV.py
# 48ms, 77.49%; 13.3MB, 35.09%
class Solution:
    def combinationSum4(self, nums: list, target: int) -> int:
        tmp = [1] + [0] * target
        for i in range(1, 1 + target):
            for n in nums:
                if i >= n: tmp[i] += tmp[i - n];
        return tmp[-1]

File: 201 - 400/377. Combination Sum IV.cc
// 4ms, 100.0%; 8.8MB, 58.06%
#include <bits/stdc++.h>
using namespace std;

static int x = []() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    return 0;
}();

class Solution {
public:
    int combinationSum4(const vector<int> &nums, int target) {
        unsigned *tmp = new unsigned[1 + target]{1};
        for (int i = 1; i <= target; ++i)
            for (auto n : nums)
                if (i >= n) tmp[i] += tmp[i - n];
        return tmp[target];
    }
};
